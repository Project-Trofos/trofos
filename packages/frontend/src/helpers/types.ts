export type PickRename<T, K extends keyof T, R extends PropertyKey> = Pick<Omit<T, K> & { [P in R]: T[K] }, R>;
