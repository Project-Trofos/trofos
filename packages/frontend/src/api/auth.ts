import trofosApiSlice from '.';

export type UserLoginInfo = {
  userEmail: string,
  userPassword: string
};

export type UserInfo = {
  userEmail: string,
  userRole: number,
  userId: number
};

export type ChangePassword = {
  userId : number | undefined,
  oldUserPassword : string,
  newUserPassword: string,
}

// Auth APIs
const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    loginUser: builder.mutation<void, UserLoginInfo>({
      query: (userLoginInfo) => ({
        url: '/account/login',
        method: 'POST',
        body : {
          userEmail: userLoginInfo.userEmail,
          userPassword: userLoginInfo.userPassword,
        },
        credentials: 'include',
      }),
    }),
    logoutUser: builder.mutation<void, void>({
      query: () => ({
        url: '/account/logout',
        method: 'POST',
        credentials: 'include',
      }),
      invalidatesTags : ['UserInfo'],
    }),
    getUserInfo: builder.query<UserInfo, void>({
      query: () => ({
        url: '/account/userInfo',
        credentials: 'include',
      }),
      providesTags: ['UserInfo'],
    }),
    changePassword : builder.mutation<void, ChangePassword>({
      query: (changePassword) => ({
        url: '/account/changePassword',
        method: 'POST',
        body : {
          userId : changePassword.userId,
          oldUserPassword : changePassword.oldUserPassword,
          newUserPassword : changePassword.newUserPassword
        },
        credentials: 'include'
      })
    })
  }),
});

export const { 
  useLoginUserMutation, 
  useLogoutUserMutation,
  useGetUserInfoQuery,
  useChangePasswordMutation
} = extendedApi;