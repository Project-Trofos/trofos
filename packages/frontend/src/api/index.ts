import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// Define a service using a base URL and expected endpoints
const trofosApiSlice = createApi({
  reducerPath: 'trofosApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NODE_ENV === 'test' ? 'http://localhost:3001/api' : '/api',
    prepareHeaders: (headers) => {
      headers.set('Cache-control', 'no-cache');
      return headers;
    },
  }),
  tagTypes: [
    'Project',
    'UserInfo',
    'Course',
    'Backlog',
    'BacklogHistory',
    'User',
    'Sprint',
    'ActionsOnRoles',
    'CourseRoles',
    'ProjectRoles',
    'Settings',
    'StandUpHeader',
    'StandUp',
    'Retrospective',
    'ProjectUserSettings',
    'Feedback',
  ],
  endpoints: () => ({}),
});

// Define a service using a base URL and expected endpoints
export const nusmodsApiSlice = createApi({
  reducerPath: 'nusmodsApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.nusmods.com/v2/2022-2023' }),
  endpoints: () => ({}),
});

export default trofosApiSlice;
