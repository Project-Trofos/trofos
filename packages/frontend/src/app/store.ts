import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
