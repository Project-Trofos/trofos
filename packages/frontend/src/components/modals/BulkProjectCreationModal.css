.bulk-project-creation-modal-list .ant-list-header {
  background-color: beige;
}
