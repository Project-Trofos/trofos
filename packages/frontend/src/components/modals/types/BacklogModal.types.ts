export interface BacklogFormFields extends FormData {
  projectId: number;
}

export type BacklogSelect = {
  id: string;
  name: string;
};
