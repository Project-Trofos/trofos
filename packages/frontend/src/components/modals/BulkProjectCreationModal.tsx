import React, { useCallback, useMemo, useState } from 'react';
import { Button, InputNumber, List, message, Modal, Space, Typography } from 'antd';
import { CourseData, ProjectData, UserData } from '../../api/types';
import { useBulkCreateProjectsMutation } from '../../api/course';
import { shuffleArray } from '../../helpers/random';

import './BulkProjectCreationModal.css';
import { UserInfo } from '../../api/auth';
import { getErrorMessage } from '../../helpers/error';

type ProjectAllocation = {
  name: string;
  users: UserData[];
}[];

/**
 * Modal for bulk creation of projects based
 */
export default function BulkProjectCreationModal({
  course,
  projects,
  currentUserInfo,
}: {
  course: CourseData | undefined;
  projects: ProjectData[] | undefined;
  currentUserInfo: UserInfo | undefined;
}) {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [allocations, setAllocations] = useState<ProjectAllocation>([]);
  const [usersPerGroup, setUsersPerGroup] = useState<number | null>(null);
  const [bulkCreateProject] = useBulkCreateProjectsMutation();

  const showModal = () => {
    setIsModalOpen(true);
  };

  const handleOk = async () => {
    try {
      if (!course || allocations.length === 0) {
        setIsModalOpen(false);
        return;
      }
      await bulkCreateProject({
        courseId: course.id,
        courseName: course.cname,
        courseYear: course.year,
        courseSem: course.sem,
        isPublic: course.public,
        projects: allocations.map((p) => ({
          projectName: p.name,
          users: p.users.map((u) => ({ userId: u.user.user_id })),
        })),
      }).unwrap();
      setIsModalOpen(false);
      message.success('Projects created!');
    } catch (e) {
      console.error(getErrorMessage(e));
      message.error('Unable to create projects! Please contact system administrator.');
    }
  };

  const handleCancel = () => {
    setIsModalOpen(false);
  };

  const usersWithoutProject: UserData[] = useMemo(() => {
    if (!course || !projects) {
      return [];
    }
    // Users without project
    const users = course.users.filter(
      (u) => !projects.some((p) => p.users.some((pu) => pu.user.user_id === u.user.user_id)),
    );
    // Current user will not be included
    return users.filter((u) => u.user.user_id !== currentUserInfo?.userId);
  }, [course, projects, currentUserInfo]);

  const generate = useCallback(() => {
    if (!course || !usersPerGroup) {
      return [];
    }
    const shuffledUsers = shuffleArray(usersWithoutProject);

    const groups: ProjectAllocation = [];
    let index = 0;
    let groupIndex = 1;
    while (index < shuffledUsers.length) {
      const users = [];
      for (let i = 0; i < usersPerGroup; i += 1) {
        if (index >= shuffledUsers.length) {
          break;
        }
        users.push(shuffledUsers[index]);
        index += 1;
      }
      groups.push({ name: `Group ${groupIndex}`, users });
      groupIndex += 1;
    }
    return groups;
  }, [course, usersPerGroup, usersWithoutProject]);

  return (
    <>
      <Button onClick={showModal}>Bulk Create</Button>
      <Modal title="Bulk Project Creation" open={isModalOpen} onOk={handleOk} onCancel={handleCancel}>
        <Space direction="vertical" style={{ width: '100%' }}>
          <Typography.Paragraph>Students without a project will be assigned a project randomly.</Typography.Paragraph>
          {usersWithoutProject.length === 0 ? (
            <Space>
              <Typography.Text>There are no students without a project!</Typography.Text>
            </Space>
          ) : (
            <>
              <Space>
                <Typography.Text>Number of students in a project:</Typography.Text>
                <InputNumber min={1} onChange={(value) => setUsersPerGroup(value)} />
                <Button onClick={() => setAllocations(generate())}>Generate</Button>
              </Space>
              {allocations.map((a, index) => (
                <List
                  className="bulk-project-creation-modal-list"
                  header={a.name}
                  bordered
                  key={a.name}
                  dataSource={a.users}
                  renderItem={(item) => <List.Item>{item.user.user_email}</List.Item>}
                />
              ))}
            </>
          )}
        </Space>
      </Modal>
    </>
  );
}
