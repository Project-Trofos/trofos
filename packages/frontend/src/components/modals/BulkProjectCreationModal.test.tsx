import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import '../../mocks/antd';

import store from '../../app/store';
import BulkProjectCreationModal from './BulkProjectCreationModal';
import { CourseData, ProjectData } from '../../api/types';
import { UserInfo } from '../../api/auth';

describe('test course creation modal', () => {
  const mockCourseData: CourseData = {
    id: 1,
    code: 'course_id',
    startYear: 2022,
    startSem: 10,
    endYear: 2022,
    endSem: 10,
    cname: 'course1',
    description: 'project1_description',
    public: false,
    created_at: '2022-09-15T01:58:01.735Z',
    milestones: [],
    announcements: [],
    shadow_course: false,
    courseRoles: [
      {
        user: {
          user_id: 1,
          user_email: 'email',
          user_display_name: 'User 1',
          courseRoles: [
            {
              id: 1,
              user_email: 'email',
              role_id: 1,
              course_id: 1,
            },
          ],
        },
      },
    ],
  };

  const mockProjectData: ProjectData[] = [];

  const mockUserInfo: UserInfo = {
    userEmail: 'email',
    userDisplayName: 'User 1',
    userId: 1,
    userRoleActions: [],
  };

  const setup = (course: CourseData | undefined, currentUserInfo: UserInfo | undefined, projects: ProjectData[]) => {
    const { baseElement, debug } = render(
      <Provider store={store}>
        <BulkProjectCreationModal course={course} currentUserInfo={currentUserInfo} projects={projects} />
      </Provider>,
    );

    // Open modal
    const button = screen.getByText(/bulk create/i);
    fireEvent.click(button);

    return { baseElement, debug };
  };

  it('should render fields if there are users without project', () => {
    // Use a different ID
    const { baseElement } = setup(mockCourseData, { ...mockUserInfo, userId: 2 }, mockProjectData);

    // Ensure fields are present
    expect(screen.getByText(/Number of students in a project:/i)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });

  it('should not render fields if there no user without a project', () => {
    setup(mockCourseData, mockUserInfo, mockProjectData);

    expect(screen.queryByText(/Number of students in a project:/i)).toBeNull();
  });

  it('should be able to generate groups', async () => {
    // Use a different ID
    setup(mockCourseData, { ...mockUserInfo, userId: 2 }, mockProjectData);

    const input = screen.getByRole('spinbutton');

    fireEvent.change(input, { target: { value: 1 } });

    const button = screen.getByText('Generate');

    fireEvent.click(button);

    await screen.findByText(/Group 1/i);
    expect(screen.getByText(mockUserInfo.userDisplayName)).toBeInTheDocument();
  });
});
