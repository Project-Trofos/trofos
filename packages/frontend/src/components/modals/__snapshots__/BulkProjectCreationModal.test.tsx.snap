// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test course creation modal > should render fields if there are users without project 1`] = `
<body>
  <div>
    <button
      class="ant-btn ant-btn-default"
      type="button"
    >
      <span>
        Bulk Create
      </span>
    </button>
  </div>
  <div>
    <div
      class="ant-modal-root"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal"
          role="dialog"
          style="width: 520px; transform-origin: 0px 0px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <button
              aria-label="Close"
              class="ant-modal-close"
              type="button"
            >
              <span
                class="ant-modal-close-x"
              >
                <span
                  aria-label="close"
                  class="anticon anticon-close ant-modal-close-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="close"
                    fill="currentColor"
                    fill-rule="evenodd"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="test-id"
              >
                Bulk Project Creation
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <div
                class="ant-space ant-space-vertical ant-space-gap-row-middle ant-space-gap-col-middle"
                style="display: flex;"
              >
                <div
                  class="ant-space-item"
                >
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
                    style="display: flex; justify-content: space-between;"
                  />
                </div>
                <div
                  class="ant-space-item"
                >
                  <div
                    class="ant-table-wrapper"
                  >
                    <div
                      class="ant-spin-nested-loading"
                    >
                      <div
                        class="ant-spin-container"
                      >
                        <div
                          class="ant-table ant-table-small ant-table-bordered"
                        >
                          <div
                            class="ant-table-container"
                          >
                            <div
                              class="ant-table-content"
                            >
                              <table
                                style="table-layout: auto;"
                              >
                                <colgroup>
                                  <col
                                    class="ant-table-selection-col"
                                  />
                                  <col
                                    style="width: 150px;"
                                  />
                                  <col />
                                  <col />
                                  <col
                                    style="width: 150px;"
                                  />
                                </colgroup>
                                <thead
                                  class="ant-table-thead"
                                >
                                  <tr>
                                    <th
                                      class="ant-table-cell ant-table-selection-column"
                                      scope="col"
                                    >
                                      <div
                                        class="ant-table-selection"
                                      >
                                        <label
                                          class="ant-checkbox-wrapper"
                                        >
                                          <span
                                            class="ant-checkbox ant-wave-target"
                                          >
                                            <input
                                              aria-label="Select all"
                                              class="ant-checkbox-input"
                                              type="checkbox"
                                            />
                                            <span
                                              class="ant-checkbox-inner"
                                            />
                                          </span>
                                        </label>
                                      </div>
                                    </th>
                                    <th
                                      aria-label="ID"
                                      class="ant-table-cell ant-table-column-has-sorters"
                                      scope="col"
                                      tabindex="0"
                                    >
                                      <div
                                        class="ant-table-column-sorters"
                                      >
                                        <span
                                          class="ant-table-column-title"
                                        >
                                          ID
                                        </span>
                                        <span
                                          class="ant-table-column-sorter ant-table-column-sorter-full"
                                        >
                                          <span
                                            aria-hidden="true"
                                            class="ant-table-column-sorter-inner"
                                          >
                                            <span
                                              aria-label="caret-up"
                                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                                              role="img"
                                            >
                                              <svg
                                                aria-hidden="true"
                                                data-icon="caret-up"
                                                fill="currentColor"
                                                focusable="false"
                                                height="1em"
                                                viewBox="0 0 1024 1024"
                                                width="1em"
                                              >
                                                <path
                                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                                />
                                              </svg>
                                            </span>
                                            <span
                                              aria-label="caret-down"
                                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                                              role="img"
                                            >
                                              <svg
                                                aria-hidden="true"
                                                data-icon="caret-down"
                                                fill="currentColor"
                                                focusable="false"
                                                height="1em"
                                                viewBox="0 0 1024 1024"
                                                width="1em"
                                              >
                                                <path
                                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                                />
                                              </svg>
                                            </span>
                                          </span>
                                        </span>
                                      </div>
                                    </th>
                                    <th
                                      aria-label="Email"
                                      class="ant-table-cell ant-table-column-has-sorters"
                                      scope="col"
                                      tabindex="0"
                                    >
                                      <div
                                        class="ant-table-column-sorters"
                                      >
                                        <span
                                          class="ant-table-column-title"
                                        >
                                          Email
                                        </span>
                                        <span
                                          class="ant-table-column-sorter ant-table-column-sorter-full"
                                        >
                                          <span
                                            aria-hidden="true"
                                            class="ant-table-column-sorter-inner"
                                          >
                                            <span
                                              aria-label="caret-up"
                                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                                              role="img"
                                            >
                                              <svg
                                                aria-hidden="true"
                                                data-icon="caret-up"
                                                fill="currentColor"
                                                focusable="false"
                                                height="1em"
                                                viewBox="0 0 1024 1024"
                                                width="1em"
                                              >
                                                <path
                                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                                />
                                              </svg>
                                            </span>
                                            <span
                                              aria-label="caret-down"
                                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                                              role="img"
                                            >
                                              <svg
                                                aria-hidden="true"
                                                data-icon="caret-down"
                                                fill="currentColor"
                                                focusable="false"
                                                height="1em"
                                                viewBox="0 0 1024 1024"
                                                width="1em"
                                              >
                                                <path
                                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                                />
                                              </svg>
                                            </span>
                                          </span>
                                        </span>
                                      </div>
                                    </th>
                                    <th
                                      class="ant-table-cell"
                                      scope="col"
                                    >
                                      Display Name
                                    </th>
                                    <th
                                      aria-label="Role"
                                      class="ant-table-cell ant-table-column-has-sorters"
                                      scope="col"
                                      tabindex="0"
                                    >
                                      <div
                                        class="ant-table-column-sorters"
                                      >
                                        <span
                                          class="ant-table-column-title"
                                        >
                                          Role
                                        </span>
                                        <span
                                          class="ant-table-column-sorter ant-table-column-sorter-full"
                                        >
                                          <span
                                            aria-hidden="true"
                                            class="ant-table-column-sorter-inner"
                                          >
                                            <span
                                              aria-label="caret-up"
                                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                                              role="img"
                                            >
                                              <svg
                                                aria-hidden="true"
                                                data-icon="caret-up"
                                                fill="currentColor"
                                                focusable="false"
                                                height="1em"
                                                viewBox="0 0 1024 1024"
                                                width="1em"
                                              >
                                                <path
                                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                                />
                                              </svg>
                                            </span>
                                            <span
                                              aria-label="caret-down"
                                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                                              role="img"
                                            >
                                              <svg
                                                aria-hidden="true"
                                                data-icon="caret-down"
                                                fill="currentColor"
                                                focusable="false"
                                                height="1em"
                                                viewBox="0 0 1024 1024"
                                                width="1em"
                                              >
                                                <path
                                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                                />
                                              </svg>
                                            </span>
                                          </span>
                                        </span>
                                      </div>
                                    </th>
                                  </tr>
                                </thead>
                                <tbody
                                  class="ant-table-tbody"
                                >
                                  <tr
                                    class="ant-table-row ant-table-row-level-0"
                                    data-row-key="1"
                                  >
                                    <td
                                      class="ant-table-cell ant-table-selection-column"
                                    >
                                      <label
                                        class="ant-checkbox-wrapper"
                                      >
                                        <span
                                          class="ant-checkbox ant-wave-target"
                                        >
                                          <input
                                            class="ant-checkbox-input"
                                            type="checkbox"
                                          />
                                          <span
                                            class="ant-checkbox-inner"
                                          />
                                        </span>
                                      </label>
                                    </td>
                                    <td
                                      class="ant-table-cell"
                                    >
                                      1
                                       
                                    </td>
                                    <td
                                      class="ant-table-cell"
                                    >
                                      email
                                       
                                    </td>
                                    <td
                                      class="ant-table-cell"
                                    >
                                      User 1
                                    </td>
                                    <td
                                      class="ant-table-cell"
                                    >
                                      STUDENT
                                    </td>
                                  </tr>
                                </tbody>
                              </table>
                            </div>
                          </div>
                          <div
                            class="ant-table-footer"
                          >
                            You have selected 0 user(s).
                          </div>
                        </div>
                        <ul
                          class="ant-pagination ant-pagination-mini ant-table-pagination ant-table-pagination-right"
                        >
                          <li
                            aria-disabled="true"
                            class="ant-pagination-prev ant-pagination-disabled"
                            title="Previous Page"
                          >
                            <button
                              class="ant-pagination-item-link"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                aria-label="left"
                                class="anticon anticon-left"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="left"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </li>
                          <li
                            class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                            tabindex="0"
                            title="1"
                          >
                            <a
                              rel="nofollow"
                            >
                              1
                            </a>
                          </li>
                          <li
                            aria-disabled="true"
                            class="ant-pagination-next ant-pagination-disabled"
                            title="Next Page"
                          >
                            <button
                              class="ant-pagination-item-link"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                aria-label="right"
                                class="anticon anticon-right"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="right"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </li>
                        </ul>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-space ant-space-vertical ant-space-gap-row-small ant-space-gap-col-small"
                style="width: 100%;"
              >
                <div
                  class="ant-space-item"
                >
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
                  >
                    <div
                      class="ant-space-item"
                    >
                      <span
                        class="ant-typography"
                      >
                        Number of students in a project:
                      </span>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div
                        class="ant-input-number ant-input-number-sm"
                      >
                        <div
                          class="ant-input-number-handler-wrap"
                        >
                          <span
                            aria-disabled="false"
                            aria-label="Increase Value"
                            class="ant-input-number-handler ant-input-number-handler-up"
                            role="button"
                            unselectable="on"
                          >
                            <span
                              aria-label="up"
                              class="anticon anticon-up ant-input-number-handler-up-inner"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                                />
                              </svg>
                            </span>
                          </span>
                          <span
                            aria-disabled="false"
                            aria-label="Decrease Value"
                            class="ant-input-number-handler ant-input-number-handler-down"
                            role="button"
                            unselectable="on"
                          >
                            <span
                              aria-label="down"
                              class="anticon anticon-down ant-input-number-handler-down-inner"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                />
                              </svg>
                            </span>
                          </span>
                        </div>
                        <div
                          class="ant-input-number-input-wrap"
                        >
                          <input
                            aria-valuemin="1"
                            autocomplete="off"
                            class="ant-input-number-input"
                            role="spinbutton"
                            step="1"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <button
                        class="ant-btn ant-btn-default ant-btn-sm"
                        type="button"
                      >
                        <span>
                          Generate
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-modal-footer"
            >
              <button
                class="ant-btn ant-btn-default"
                type="button"
              >
                <span>
                  Cancel
                </span>
              </button>
              <button
                class="ant-btn ant-btn-primary"
                type="button"
              >
                <span>
                  OK
                </span>
              </button>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
